File: cpp/tree/tree.cpp
#include "tree.hpp"

template <typename T>
node<T> *tree<T>::insert(node<T> *root, T value)
{
  if (root == nullptr)
  {
    return new node(value);
    size++;
  }
  if (value < root->data)
  {
    root->left = insert(root->left, value);
  }
  else if (value > root->data)
  {
    root->right = insert(root->right, value);
  }
  return root;
}

template <typename T>
node<T> *tree<T>::del(node<T> *root, T value)
{
  if (root == nullptr){
    return nullptr;
  }
  if (value < root->data){
    root->left = removeHelper(root->left, value);
  }
  else if (value > root->data){
    root->right = removeHelper(root->right, value);
  }
  else{
    if (root->left == nullptr){
      node<T> *temp = root->right;
      delete root;
      return temp;
    }
    else if (root->right == nullptr){
      node<T> *temp = root->left;
      delete root;
      return temp;
    }
    node<T> *temp = findMin(root->right);
    root->data = temp->data;
    root->right = removeHelper(root->right, temp->data);
  }
  return root;
}

template <typename T>
node<T> tree<T>::*findMin(node<T> *root) {
  while (root->left != nullptr){
    root = root->left;
  }
  return root;
}

template <typename T>
bool tree<T>::isRootExist(node<T> *root, T value)
{
  try
  {
    // If node Null
    if (root == nullptr)
      return false;
    // If he find it
    if (value == root->data)
      return true;
    // here he call him self
    if (value < root->data)
      return searchHelper(root->left, value);
    else
      return searchHelper(root->right, value);
  }
  catch (const std::exception &e)
  {
    return false;
    std::cerr << e.what() << '\n';
  }
}

template <typename T>
node<T> *tree<T>::searchFor(node<T> *root, T value)
{
  try
  {
    // If node Null
    if (root == nullptr)
      return root;
    // If he find it
    if (value == root->data)
      return root;
    // here he call him self
    if (value < root->data)
      return searchHelper(root->left, value);
    else
      return searchHelper(root->right, value);
  }
  catch (const std::exception &e)
  {
    std::cerr << e.what() << '\n';
  }
}

template <typename T>
void tree<T>::preorder(node<T> *root)
{
  if (root != nullptr)
  {
    std::cout << root->data << " ";
    preorder(root->left);
    preorder(root->right);
  }
}

template <typename T>
void tree<T>::inorder(node<T> *root)
{
  if (root != nullptr)
  {
    inorder(root->left);
    std::cout << root->data << " ";
    inorder(root->right);
  }
}

template <typename T>
void tree<T>::postorder(node<T> *root)
{
  if (root != nullptr)
  {
    postorder(root->left);
    postorder(root->right);
    std::cout << root->data << " ";
  }
}

template <typename T>
bool tree<T>::insert(T value)
{
  try{
    mainRoot = insert(mainRoot, value);
    return true;
  }
  catch (const std::exception &e)
  {
    std::cerr << e.what() << '\n';
    return false;
  }
}

template <typename T>
bool tree<T>::del(T value){
  try{
    mainRoot = del(mainRoot, value);
    return true;
  }
  catch(const std::exception& e){
    std::cerr << e.what() << '\n';
  }
}

template <typename T>
void tree<T>::preorder(){
  cout << "Pre-order: ";
  preorder(mainRoot);
  cout << endl;
}

template <typename T>
void tree<T>::inorder()
{
  cout << "In-order: ";
  inorder(mainRoot);
  cout << endl;
}

template <typename T>
void tree<T>::postorder(){
  cout << "Post-order: ";
  postorder(mainRoot);
  cout << endl;
}

template <typename T>
bool tree<T>::isRootExist(T value){
  try{
    return isRootExist(mainRoot, value);
  }
  catch(const std::exception& e){
    return false;
  }
}

// find min
template <typename T>
node<T> tree<T>::findMin(){
  return findMin(mainRoot);
}
// Search for an elements helper
template <typename T>
node<T> tree<T>::searchFor(T value){
  try{
    return searchFor(mainRoot, value);
  }
  catch(const std::exception& e){
    std::cerr << e.what() << '\n';
    return node<T>();
  }
}

File: python/tree/tree.py
class Node:
    def __init__(self, value):
        self.data = value
        self.left = None
        self.right = None

class Tree:
    def __init__(self):
        self.size = 0
        self.main_root = None

    def _insert(self, root, value):
        if root is None:
            return Node(value)
        if value < root.data:
            root.left = self._insert(root.left, value)
        else:
            root.right = self._insert(root.right, value)
        return root

    def insert(self, value):
        self.main_root = self._insert(self.main_root, value)
        self.size += 1
        return True

    def _find_min(self, root):
        while root.left is not None:
            root = root.left
        return root

    def _delete(self, root, value):
        if root is None:
            return root
        if value < root.data:
            root.left = self._delete(root.left, value)
        elif value > root.data:
            root.right = self._delete(root.right, value)
        else:
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left
            temp = self._find_min(root.right)
            root.data = temp.data
            root.right = self._delete(root.right, temp.data)
        return root

    def delete(self, value):
        self.main_root = self._delete(self.main_root, value)
        self.size -= 1
        return True

    def _is_root_exist(self, root, value):
        if root is None:
            return False
        if root.data == value:
            return True
        elif value < root.data:
            return self._is_root_exist(root.left, value)
        else:
            return self._is_root_exist(root.right, value)

    def is_root_exist(self, value):
        return self._is_root_exist(self.main_root, value)

    def _search_for(self, root, value):
        if root is None or root.data == value:
            return root
        if value < root.data:
            return self._search_for(root.left, value)
        return self._search_for(root.right, value)

    def search_for(self, value):
        return self._search_for(self.main_root, value)

    def _preorder(self, root):
        if root:
            print(root.data, end=" ")
            self._preorder(root.left)
            self._preorder(root.right)

    def _inorder(self, root):
        if root:
            self._inorder(root.left)
            print(root.data, end=" ")
            self._inorder(root.right)

    def _postorder(self, root):
        if root:
            self._postorder(root.left)
            self._postorder(root.right)
            print(root.data, end=" ")

    def preorder(self):
        self._preorder(self.main_root)
        print()

    def inorder(self):
        self._inorder(self.main_root)
        print()

    def postorder(self):
        self._postorder(self.main_root)
        print()

    def find_min(self):
        return self._find_min(self.main_root).data if self.main_root else None

